File: Languages/python/algorithms-dsa/leetcode/trees/invert_binary_tree.py
"""
LeetCode 226: Invert Binary Tree
Given the root of a binary tree, invert the tree and return its root.

Time: O(n), Space: O(h) where h is height
"""

from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if root is None:
        return None
    root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root


def tree_to_list(root: Optional[TreeNode]) -> list:
    """Level-order traversal for testing."""
    if not root:
        return []
    result = []
    queue = [root]
    while queue:
        node = queue.pop(0)
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    # Remove trailing Nones
    while result and result[-1] is None:
        result.pop()
    return result


if __name__ == "__main__":
    # Tree: [4, 2, 7, 1, 3, 6, 9]
    root = TreeNode(4,
        TreeNode(2, TreeNode(1), TreeNode(3)),
        TreeNode(7, TreeNode(6), TreeNode(9))
    )
    inverted = invert_tree(root)
    assert tree_to_list(inverted) == [4, 7, 2, 9, 6, 3, 1]

    # Empty tree
    assert invert_tree(None) is None

    # Single node
    single = TreeNode(1)
    assert tree_to_list(invert_tree(single)) == [1]

    print("All tests passed!")

File: Languages/java/algorithms-dsa/leetcode/trees/InvertBinaryTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * LeetCode 226: Invert Binary Tree
 * Given the root of a binary tree, invert the tree and return its root.
 * 
 * Time: O(n), Space: O(h)
 */
public class InvertBinaryTree {

    static class TreeNode {
        int val;
        TreeNode left, right;

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public static TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        TreeNode temp = root.left;
        root.left = invertTree(root.right);
        root.right = invertTree(temp);
        return root;
    }

    public static List<Integer> levelOrder(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            result.add(node.val);
            if (node.left != null) queue.add(node.left);
            if (node.right != null) queue.add(node.right);
        }
        return result;
    }

    public static void main(String[] args) {
        // Tree: [4, 2, 7, 1, 3, 6, 9]
        TreeNode root = new TreeNode(4,
            new TreeNode(2, new TreeNode(1), new TreeNode(3)),
            new TreeNode(7, new TreeNode(6), new TreeNode(9))
        );

        TreeNode inverted = invertTree(root);
        List<Integer> result = levelOrder(inverted);
        List<Integer> expected = List.of(4, 7, 2, 9, 6, 3, 1);
        assert result.equals(expected) : "Tree inversion failed: " + result;

        // Null tree
        assert invertTree(null) == null : "Null tree test failed";

        System.out.println("All tests passed!");
    }
}
